llvm/lib/Transforms/IPO/RepoPruning.cpp

```diff
@@ -325,23 +325,41 @@ bool RepoPruning::runOnModule(Module &M) {
   }
 
   // Enable this pass support for alias.
-  for (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
-    Module::alias_iterator CurI = I;
-    ++I;
+  // First, check whether a GlobalAlias references a pruned GO. If yes, record
+  // it. Second, replace all recorded aliases by the declaration. These two
+  // steps can't be merged since the second step could change the base object of
```

Comment:
I don’t really understand your explanation here. You finish by saying that it will “result in the backend error” but don’t say what causes the error.

Reply:
Thanks, the comment has been modified.

Reply:
Do you mean that changing the base object of the alias will cause an (unspecified) backend error? How does splitting the algorithm into two steps avoid this problem?

Reply:
> 
> 
> Do you mean that changing the base object of the alias will cause an (unspecified) backend error? 

Yes, thinking about the following IR code after RepoMetadataGenerationPass.
```
; ModuleID = 'alias.cpp'
target triple = "x86_64-pc-linux-gnu-repo"

%class.base = type { i8 }
%class.derived = type { i8 }

@_ZN4baseD1Ev = unnamed_addr alias void (%class.base*), void (%class.base*)* @_ZN4baseD2Ev
@_ZN7derivedD2Ev = unnamed_addr alias void (%class.derived*), bitcast (void (%class.base*)* @_ZN4baseD2Ev to void (%class.derived*)*)
@_ZN7derivedD1Ev = unnamed_addr alias void (%class.derived*), void (%class.derived*)* @_ZN7derivedD2Ev

; Function Attrs: nounwind uwtable
define void @_ZN4baseD2Ev(%class.base* %this) unnamed_addr #0 align 2 !repo_ticket !2 {
entry:
  ret void
}
```
There are 3 aliases in the Module and follow the order: 1) _ZN4baseD1Ev 2) _ZN7derivedD2Ev 3) _ZN7derivedD1Ev. The initial algorithm:
```
  For (Module::alias_iterator I = M.alias_begin(), E = M.alias_end(); I != E;) {
    Module::alias_iterator CurI = I;
    ++I;
    if (auto *Aliasee = CurI->getBaseObject()) {
      if (wasPruned(*Aliasee)) {
        Type *Ty = CurI->getValueType();
        CurI->removeFromParent();
        CurI->replaceAllUsesWith(createDeclararion(M, Ty, CurI));
        delete &*CurI;
        ++NumAliases;
      }
    }
  }
```

- Iteration 1:  `_ZN4baseD1Ev`: 

Its base object is the `_ZN4baseD2Ev`, which is a pruned global object. `_ZN4baseD1Ev` is removed and changed from definition to declaration.

- Iteration 2:  `_ZN7derivedD2Ev`: 

Its base object is the `_ZN4baseD2Ev`, which is a pruned global object. `_ZN7derivedD2Ev ` is removed and changed from definition to declaration.

- Iteration 3:  `_ZN7derivedD1Ev `:

Without Iteration 2, its base object is the `_ZN4baseD2Ev`. However, iteration 2 has changed `_ZN7derivedD2Ev ` to the declaration. Therefore, the base object of `_ZN7derivedD1Ev ` is changed from `_ZN4baseD2Ev` to `_ZN7derivedD2Ev` (This is wrong). Since `_ZN7derivedD2Ev` is not a global object (it is an alias), the `_ZN7derivedD1Ev ` is not removed.

The IR code after the pruning pass is:
```
; ModuleID = 'alias.cpp'
source_filename = "alias.cpp"
target datalayout = "e-m:r-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-pc-linux-gnu-repo"

%class.derived = type { i8 }
%class.base = type { i8 }

@_ZN7derivedD1Ev = unnamed_addr alias void (%class.derived*), void (%class.derived*)* @_ZN7derivedD2Ev

; Function Attrs: nounwind uwtable
define available_externally void @_ZN4baseD2Ev(%class.base* %this) unnamed_addr #0 align 2 !repo_ticket !2 {
entry:
  ret void
}

declare void @_ZN4baseD1Ev(%class.base*)
declare void @_ZN7derivedD2Ev(%class.derived*)
…
```
`_ZN7derivedD1Ev` must reference to a definition, but `_ZN7derivedD2Ev` is a declaration. This results in a backend error.


>
>How does splitting the algorithm into two steps avoid this problem?
>
When separating the algorithm into two steps, the base object of `_ZN7derivedD1Ev` will keep correct.
Step1: this step collects Aliases, which will be removed in the step2.

- Iteration 1:  `_ZN4baseD1Ev` is put into the `Aliases`.

- Iteration 2:  `_ZN7derivedD2Ev` is put into the `Aliases`. 

- Iteration 3:  `_ZN7derivedD1Ev `:

Since `_ZN7derivedD2Ev` is not removed, the base object of `_ZN7derivedD1Ev `is the `_ZN4baseD2Ev`, which is a pruned global object. `_ZN7derivedD1Ev ` is put into the `Aliases`.

Step 2: Removed all aliases in the `Aliases`.
`_ZN7derivedD1Ev ` is correctly changed to the declaration. There is no backend error.